javatests/arcs/android/storage/service/StorageServiceNgImplTest.kt

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.android.storage.service
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import arcs.android.crdt.ParcelableCrdtType
+import arcs.android.storage.service.testing.FakeMessageCallback
+import arcs.android.storage.service.testing.FakeStorageChannelCallback
+import arcs.android.storage.toParcelToByteArray
+import arcs.core.data.CountType
+import arcs.core.storage.FixedDriverFactory
+import arcs.core.storage.StorageKey
+import arcs.core.storage.StorageKeyParser
+import arcs.core.storage.StoreOptions
+import arcs.core.storage.UntypedProxyMessage
+import arcs.core.storage.keys.RamDiskStorageKey
+import arcs.core.storage.testutil.MockDriverProvider
+import arcs.core.storage.testutil.testWriteBackProvider
+import arcs.core.type.Type
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.test.TestCoroutineScope
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class StorageServiceNgImplTest {
+
+  private val driverFactory = FixedDriverFactory(MockDriverProvider())
+  private val onProxyMessageCallback: suspend (StorageKey, UntypedProxyMessage) -> Unit =
+    { _: StorageKey, _: UntypedProxyMessage -> }
+  private val messageCallback = FakeMessageCallback()
+  private val channelCallback = FakeStorageChannelCallback()
+
+  private val scope = TestCoroutineScope()
+  private lateinit var storageService: StorageServiceNgImpl
+
+  @Before
+  fun setUp() {
+    StorageKeyParser.addParser(RamDiskStorageKey)
+    storageService = StorageServiceNgImpl(
+      scope,
+      driverFactory,
+      ::testWriteBackProvider,
+      null,
+      onProxyMessageCallback
+    )
+  }
+
+  @After
+  fun tearDown() {
+    scope.cleanupTestCoroutines()
+  }
+
+  @Test
+  fun openStorageChannel_invokesOnCreateCallback_andCreatesChannel() = scope.runBlockingTest {
+    val encodedStoreOptions = createEncodedStoreOptions(
+      RamDiskStorageKey("myCount"),
```

Comment:
Optional: you could define a constant for this, in a companion object at the bottom of the file. e.g. `DUMMY_STORAGE_KEY` or something like that. You can use that in any test where you don't really care about the specifics of the storage key. Whereas a test like `openingTwoChannels_forDifferentStores_createsTwoStores` does care about the different storage keys, so you wouldn't use it there.

You could also define `DUMMY_STORE_OPTIONS`, that would make things even shorter. Again, you can use that anywhere you like, but not in a test like `openStorageChannel_createsAStore`, since it asserts on the specific storage key that's encoded inside the `StoreOptions`, and you shouldn't rely on a constant like `DUMMY_STORE_OPTIONS` having its storage key set to any particular value.

Reply:
Creating `DUMMY_STORE_OPTIONS` would either require using `createEncodedStoreOptions` or duplicating the logic in `createEncodedStoreOptions`. What's the best way to create `DUMMY_STORE_OPTIONS` in the companion object? I was struggling to find an elegant way to use `createEncodedStoreOptions`.

Reply:
`createEncodedStoreOptions` doesn't use any members from the class, so it can go inside the companion object too (i.e. can be static).

You're free to call other methods from the companion object within it:

```
private companion object {
  private val DUMMY_STORE_OPTIONS = createEncodedStoreOptions(...)

  private fun createEncodedStoreOptions(...) {
      return ...
  }
}
```